pennylane_lightning/core/jaxpr_adjoint.py

```diff
@@ -0,0 +1,123 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""This module contains utility functions for working with JAX Jaxprs in the context of adjoint differentiation."""
+
+from typing import Sequence, Tuple
+
+import jax
+import numpy as np
+import pennylane as qml
+from pennylane.capture.primitives import AbstractMeasurement
+from pennylane.typing import TensorLike
+
+
+def validate_args_tangents(
+    args: Sequence[TensorLike], tangents: Sequence[TensorLike]
+) -> Tuple[TensorLike]:
+    """Ensure args and tangents provided to the JAXPR are consistent.
+
+    Args:
+        args (Sequence[TensorLike]): the arguments to the JAXPR
+        tangents (Sequence[TensorLike]): the tangents to the JAXPR
+
+    Returns:
+        Tuple[TensorLike]: the validated tangents
+    """
+
+    if len(args) != len(tangents):
+        raise ValueError("The number of arguments and tangents must match")
+
+    def _make_zero(tan, arg):
+        return jax.lax.zeros_like_array(arg) if isinstance(tan, jax.interpreters.ad.Zero) else tan
+
+    tangents = tuple(map(_make_zero, tangents, args))
+
+    for tan in tangents:
+        if jax.numpy.issubdtype(jax.numpy.asarray(tan).dtype, jax.numpy.integer):
+            # will be able to support ad.Zero integer tangents when can track trainable parameters
+            raise ValueError("Tangents cannot be of integer type yet.")
+
+    return tangents
+
+
+def get_output_shapes(jaxpr: "jax.core.Jaxpr", num_wires: int) -> Tuple:
+    """
+    Compute the output shapes of the JAXPR and the Jacobian shapes of the JAXPR.
+
+    Args:
+        jaxpr (jax.core.Jaxpr): the JAXPR to analyze
+        num_wires (int): the number of wires
+
+    Returns:
+        Tuple: a tuple containing the output shapes of the JAXPR and the Jacobian shapes of the JAXPR
+    """
+
+    # pylint: disable=no-member
+    dtype_map = {
+        float: (jax.numpy.float64 if jax.config.jax_enable_x64 else jax.numpy.float32),
+        int: jax.numpy.int64 if jax.config.jax_enable_x64 else jax.numpy.int32,
+        complex: (jax.numpy.complex128 if jax.config.jax_enable_x64 else jax.numpy.complex64),
+    }
+
+    def _get_shape(var):
+        if isinstance(var.aval, AbstractMeasurement):
+            s, dtype = var.aval.abstract_eval(num_device_wires=num_wires, shots=0)
+            return jax.core.ShapedArray(s, dtype_map[dtype])
+        raise NotImplementedError("The circuit should return a measurement")
+
+    def _flatten_shaped_array(aval):
+        if aval.shape == ():
+            return [aval]
+        num_elements = int(np.prod(aval.shape))
+        return [jax.core.ShapedArray((), aval.dtype) for _ in range(num_elements)]
+
+    def _get_jacobian_shape(shape_res):
+        train_args = jaxpr.invars + jaxpr.constvars
```

Comment:
So we automatically differentiate with respect to, say, closed-over variables? :thinking: Or am I misunderstanding the role of `constvars`?

Reply:
Ah, I saw in the `lightning_base` file that this is one of the restrictions, right?
Maybe good to comment very briefly on this here :)